### PR TITLE
Feat/add ssd flavor adr

### PR DESCRIPTION
**Pull request by garloff**
The ADR for adding 2 mandatory flavors with local SSD/NVMe for supporting etcd (k8s control plane node) and databases well.
Once we approve this, we will need to adjust the flavor naming standard accordingly.

### CODE REVIEW

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (fkr) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:28]: I'd refrain from naming explicit products, if not really needed.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (fkr) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:31]: In this case ("ceph in the SCS reference implementation") it makes sense to explicitly name it. (see comment above)

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (fkr) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:103]: This change gives no guarantee -- occasional write latencies

(reads imho better than the silver bullet wording)

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (fkr) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:113]: "The etcd process also gets its IO priority"

**Review by fkr — COMMENTED** [review]
Left a few small comments.
I like it - the "Motivation" part is too long for my taste. A lot of the items explained there are (imho) the category "general education for system engineers" and if we do this for every of these documents, we repeat ourselves a lot. But I might be biased and it is really helpful to have that here. 
The reason why I'd prefer a less verbose style: I fear that only few will actually read it.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:6]: There are dashes missing in this line and in the first line.

Adding them enables the Markdown renderer to understand them. Like in SCS-0001: https://raw.githubusercontent.com/SovereignCloudStack/Docs/main/Standards/scs-0001-v1-sovereign-cloud-standards.md

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Ack. I had the had hdfs in mind where a networked disk is a particularly bad match. I change the wording to "big data filesystem."

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md:31]: Thanks for clarifying.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Fixed.

**Comment by garloff** [comment]
> Left a few small comments. I like it, very well written! The "Motivation" part is a tad too long for my taste.
Obviously, I was trying to create a positive example how to write such docs.
Being more concise in the Motivation part is probably something we do want. I'll have to re-read the doc and see if I see obvious ways to do it.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md:7]: Good catch, thanks. Hopefully it's correct now.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:19]: Personally, I'd avoid categorising use cases in this much detail, as does not add too much value to the introduction and may be disputable.

Especially `gen1 (VM-based) cloud-native` strongly implies that Kubernetes is "gen2" cloud native, which might be correct, but definitely is kind of opinionated.
For example, someone might opt to avoid Kubernetes and instead build a cloud native setup by fully embracing the Hashicorp stack which in many ways kind of competes with Kubernetes: Packer for building images, Terraform for VM/resource provisioning, Nomad for workload scheduling, Consul Connect for service meshing etc.
They might not agree that this setup is a "gen1" (hence kind of legacy) setup, it's just different.

Of course, in my impression, using Kubernetes (or Kubernetes with some Hashicorp tools integrated) is the more popular choice - and also the choice SCS settled on as primary use case. Nevertheless, it might be better to focus only on the technological terms, here.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:31]:
```suggestion
flavors that use local storage. For certain workloads such as big data filesystems
or databases, local storage is highly desirable as replication may be handled at
```

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:83]: I remember that in a team call, this was a topic of discussion.

Regarding the `this would actually work, as long as at least one cluster member stays alive` bit:

According to my own experiences, and most importantly, answers of contributors in the etcd issues (https://github.com/etcd-io/etcd/issues/7445), this is not really all that easy.

According to this information, this would not work unless shut-down etcd members are manually removed, which I do not think is the default.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:110]:
```suggestion
[k8s-cluster-api-provider](https://etcd.io/docs/v3.5/op-guide/hardware/#example-hardware-configurations)
```

I guess this should be a markdown link.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:222]:
```suggestion
live-migration is supported will be subject to a flavor-metadata spec that
```

**Review by joshmue — DISMISSED** [review]
I like the "design consideration" section in particular: It is detailed, but concise.

**Review by matfechner — APPROVED** [review]
LGTM with the reguested changes

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (itrich) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:6]: According to [SCS-0001#versioning](https://github.com/SovereignCloudStack/Docs/blob/main/Standards/scs-0001-v1-sovereign-cloud-standards.md#versioning), I'd advise to drop the version number within the frontmatter of this document. Minor changes (e.g. rewording or typo fixing) should be done without touching the version number. Major changes can be indicating by increasing the version number in the filename.

**Comment by itrich** [comment]
Please don't forget to give this file a reasonable filename according to [SCS-0001#document-format](https://github.com/SovereignCloudStack/Docs/blob/main/Standards/scs-0001-v1-sovereign-cloud-standards.md#document-format) before merging this PR 😇

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (itrich) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:4]: To get used to our own standard [SCS-0001](https://github.com/SovereignCloudStack/Docs/blob/main/Standards/scs-0001-v1-sovereign-cloud-standards.md#document-format), this should be in status `Proposal`. Only the very last commit before merging should
- rename the file
- bump the status to `Draft`

**Comment by neuroserve** [comment]
What about pruning volumes from local SSDs in order to make sure, that tenants cannot access previous users data (or parts thereof) from that local SSD.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md:19]: Well, the discussion tends to be different: The k8s community considers k8s as the one and only "cloud-native" approach, which is untrue, IMVHO. Cloud-native existed before containers were widespread, but the k8s folks want to monopolize the wording. They even created a foundation with that claim.
So the wording acknowledges that a lot of current ("gen-2") cloud-native development targets containers, while other and older approaches exist ("gen-1") and continue to be valid and used.
There is no rating involved here, merely the observation that we have seen two big waves in the industry here ... the 2nd is larger and more modern.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: I would probably reverse the order, to avoid that the "big data" qualifier that is applied to filesystems also appears to apply to databases, which would be unintentional.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md:222]: Yes, thanks.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [inline Decisions/scs-xxxx-v1-ssd-flavors.md:31]:
```suggestion
flavors that use local storage. For certain workloads such as databases or big data
filesystems, local storage is highly desirable as replication may be handled at
```

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: > I would probably reverse the order, to avoid that the "big data" qualifier that is applied to filesystems also appears to apply to databases, which would be unintentional.

Done here: https://github.com/SovereignCloudStack/Docs/pull/150#discussion_r1032434391

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [reply Decisions/scs-xxxx-v1-ssd-flavors.md:31]: Continuation of https://github.com/SovereignCloudStack/Docs/pull/150#discussion_r1023737025

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (joshmue) [reply Decisions/scs-xxxx-v1-ssd-flavors.md:19]: Ah, ok. Over-interpreted the "gen" as rating, then :). Still, even when "gen" only is meant to be understood as hint regarding temporal order or history, such high level categorization in waves/generations still opens a lot of room for potential misunderstandings/discussion.

For example, Nomad was released first one year *after* Kubernetes' first release and is still advertised as alternative. Should it be considered gen-1 cloud-native, even it is "younger" than gen-2? Or gen-3?
What does gen-1 mean exactly in contrast to classical VM workload?

So, I think referring only to well-established high level terms like IaaS/... as well as concrete technical terms like VM/container would eliminate any such potential for misunderstandings or discussions unrelated to the actual topic of the decision record (like the one I just started).

**Comment by garloff** [comment]
> What about pruning volumes from local SSDs in order to make sure, that tenants cannot access previous users data (or parts thereof) from that local SSD? Shouldn't that be covered in a separate paragraph?

Good point, the volumes need to be zeroed out before being handed out to a user.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Adjusted

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: OK, ok, I used the words wave 1 and wave 2 instead.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Alright, I improved the wording, mentioning that this could be made work and that proper care needs to be taken to remove shut-down nodes.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Question is whether we want to allow for minor version updates, if we e.g. do editorial changes that clarifies the wording ... and do this in the version header without bumping the major version (which may carry new, in the worst case even incompatible changes). It's still good to reference the precise version number - which may be better to parse for humans that git hashes.

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: You never know which is the last commit before merging, no?
So as soon as the original author thinks that this is ready, can't she assign the status to Draft which becomes then effective as soon as the PR is merged?

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Changed the order of words to databases or big data filesystems, avoiding "big data" also potentially being interpreted as qualifier to databases.

**Comment by garloff** [comment]
> Good point, the volumes need to be zeroed out before being handed out to a user.

Done:

Like with networked storage, the provider must ensure that data
from previous users is not accessible (e.g. by securely erasing it
or by using a different encryption key) when local storage gets
allocated to a new VM.

**Comment by garloff** [comment]
Alright, I wanted to resolve a conversation (the one about purging volumes before reuse).
And accidentially closed the PR. Reopened now.
We need fresh approvals for being able to merge.

**Review by joshmue — APPROVED** [review]

File: Decisions/scs-xxxx-v1-ssd-flavors.md
Comment (fkr) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: @garloff *if* we use a version number, we should use semver imho.

Comment (garloff) [reply Decisions/scs-xxxx-v1-ssd-flavors.md]: Fair comment. So major.minor.patch with patches probably being clarifications, while minor versions might introduce new backwards compatible requirements (i.e. someone fulfilling v1.8.x automatically fulfills also v1.y.z with y<=7) and major bumps even possibly having requirements that conflict with earlier versions?
If we do so, the filename should probably encode major.minor, no?

**Comment by garloff** [comment]
Any chance to get an approving review, @itrich , @fkr, @matfechner, @berendt, @joshmue so we can merge this as a *draft* standard?

**Review by joshmue — APPROVED** [review]

**Pull request merged**
